File: examples/lsp_demo.py
#!/usr/bin/env python3
"""
Example: Language Server Protocol (LSP) Demo

This example demonstrates various Python code patterns that would be
handled by Depyler's LSP server for IDE integration features like:
- Code completion
- Hover information
- Go to definition
- Find references
- Diagnostics
"""

# Import statements for completion suggestions
import math
import json
from typing import List, Dict, Optional

# Global variables for reference tracking
CONFIG_FILE = "config.json"
MAX_RETRIES = 3
DEBUG_MODE = True

# Type annotations for hover information
class User:
    """Represents a user in the system."""
    
    def __init__(self, name: str, age: int):
        """Initialize a new user.
        
        Args:
            name: The user's name
            age: The user's age
        """
        self.name = name
        self.age = age
        self.created_at = None
    
    def greet(self) -> str:
        """Return a greeting message."""
        return f"Hello, I'm {self.name}!"
    
    def is_adult(self) -> bool:
        """Check if the user is an adult (18+)."""
        return self.age >= 18

# Function with rich documentation
def calculate_fibonacci(n: int) -> int:
    """Calculate the nth Fibonacci number.
    
    The Fibonacci sequence is defined as:
    - F(0) = 0
    - F(1) = 1
    - F(n) = F(n-1) + F(n-2) for n > 1
    
    Args:
        n: The position in the Fibonacci sequence
        
    Returns:
        The nth Fibonacci number
        
    Raises:
        ValueError: If n is negative
        
    Examples:
        >>> calculate_fibonacci(0)
        0
        >>> calculate_fibonacci(1)
        1
        >>> calculate_fibonacci(10)
        55
    """
    if n < 0:
        raise ValueError("n must be non-negative")
    
    if n <= 1:
        return n
    
    # Variable references for tracking
    prev, curr = 0, 1
    for _ in range(2, n + 1):
        prev, curr = curr, prev + curr
    
    return curr

# Complex function for go-to-definition
def process_users(users: List[User], filter_adults: bool = True) -> Dict[str, List[User]]:
    """Process and categorize users.
    
    Args:
        users: List of User objects to process
        filter_adults: Whether to filter only adult users
        
    Returns:
        Dictionary with 'adults' and 'minors' keys
    """
    result = {"adults": [], "minors": []}
    
    for user in users:
        # Method call - LSP should resolve User.is_adult()
        if user.is_adult():
            result["adults"].append(user)
        else:
            result["minors"].append(user)
    
    # Using global variable - LSP should find definition
    if DEBUG_MODE:
        print(f"Processed {len(users)} users")
    
    return result

# Nested function definitions
def create_handler(prefix: str):
    """Create a message handler with a prefix."""
    
    def handler(message: str) -> str:
        """Handle a message by adding prefix."""
        # Reference to outer scope variable
        return f"{prefix}: {message}"
    
    return handler

# Class with inheritance for symbol hierarchy
class AdminUser(User):
    """An admin user with special privileges."""
    
    def __init__(self, name: str, age: int, permissions: List[str]):
        super().__init__(name, age)
        self.permissions = permissions
    
    def has_permission(self, permission: str) -> bool:
        """Check if admin has a specific permission."""
        return permission in self.permissions
    
    def greet(self) -> str:
        """Admin-specific greeting."""
        # Overrides parent method
        return f"Hello, I'm Admin {self.name}!"

# Function with multiple references
def validate_age(age: int) -> bool:
    """Validate if age is in acceptable range."""
    MIN_AGE = 0
    MAX_AGE = 150
    
    # Multiple references to parameters and local variables
    if age < MIN_AGE:
        return False
    if age > MAX_AGE:
        return False
    
    # Reference to the parameter again
    return age >= 0

# Code with potential diagnostics
def problematic_function():
    """Function that might trigger LSP diagnostics."""
    
    # Unused variable (warning)
    unused_var = 42
    
    # Type mismatch (error in typed context)
    result: int = "not an int"  # Type error
    
    # Undefined variable reference (error)
    # print(undefined_variable)
    
    # Unreachable code (warning)
    return result
    print("This won't execute")

# Lambda and comprehension expressions
data_processors = {
    "double": lambda x: x * 2,
    "square": lambda x: x ** 2,
    "stringify": lambda x: str(x)
}

# List comprehension with references
def process_numbers(numbers: List[int]) -> List[int]:
    """Process numbers using various transformations."""
    # Reference to data_processors
    doubled = [data_processors["double"](n) for n in numbers]
    
    # Nested comprehension
    matrix = [[i * j for j in range(3)] for i in range(3)]
    
    # Generator expression
    sum_of_squares = sum(n ** 2 for n in numbers if n > 0)
    
    return doubled

# Decorator usage
def log_calls(func):
    """Decorator to log function calls."""
    def wrapper(*args, **kwargs):
        print(f"Calling {func.__name__}")
        return func(*args, **kwargs)
    return wrapper

@log_calls
def important_operation(value: str) -> str:
    """An important operation that should be logged."""
    return value.upper()

# Context manager for with statements
class FileManager:
    """Context manager for file operations."""
    
    def __init__(self, filename: str, mode: str = 'r'):
        self.filename = filename
        self.mode = mode
        self.file = None
    
    def __enter__(self):
        self.file = open(self.filename, self.mode)
        return self.file
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.file:
            self.file.close()

# Main execution with various LSP triggers
if __name__ == "__main__":
    # Create users - completion should suggest User class
    users = [
        User("Alice", 25),
        User("Bob", 17),
        AdminUser("Charlie", 30, ["read", "write", "delete"])
    ]
    
    # Process users - hover should show function signature
    categorized = process_users(users)
    
    # Access results - completion should suggest dict keys
    adults = categorized["adults"]
    
    # Method calls - go-to-definition should work
    for user in adults:
        print(user.greet())
    
    # Using imported modules - completion should work
    pi_value = math.pi
    sqrt_2 = math.sqrt(2)
    
    # JSON operations - hover should show types
    config_data = {"debug": DEBUG_MODE, "retries": MAX_RETRIES}
    json_str = json.dumps(config_data)
    
    # Fibonacci calculation - references should be found
    fib_10 = calculate_fibonacci(10)
    print(f"10th Fibonacci number: {fib_10}")
    
    # Handler creation - nested scopes
    info_handler = create_handler("INFO")
    error_handler = create_handler("ERROR")
    
    print(info_handler("Application started"))
    print(error_handler("Something went wrong"))
    
    # File operations with context manager
    try:
        with FileManager(CONFIG_FILE, 'w') as f:
            f.write(json_str)
    except IOError as e:
        print(f"Error writing config: {e}")
    
    # Data processing
    numbers = [1, 2, 3, 4, 5]
    processed = process_numbers(numbers)
    print(f"Processed: {processed}")
    
    # Decorator in action
    result = important_operation("hello world")
    print(f"Result: {result}")

File: examples/lsp_demo.rs
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(unreachable_patterns)]
#![allow(unused_assignments)]
#![allow(dead_code)]
use std::f64 as math;
pub const CONFIG_FILE: &str = "config.json";
pub const MAX_RETRIES: i32 = 3;
pub const DEBUG_MODE: bool = true;
pub static data_processors: std::sync::LazyLock<std::collections::HashMap<String, String>> =
    std::sync::LazyLock::new(|| {
        let mut map = HashMap::new();
        map.insert("double".to_string().to_string(), move |x| x * 2);
        map.insert("square".to_string().to_string(), move |x| {
            if 2 >= 0 && (2 as i64) <= (u32::MAX as i64) {
                ({ x } as i32)
                    .checked_pow({ 2 } as u32)
                    .expect("Power operation overflowed")
            } else {
                ({ x } as f64).powf({ 2 } as f64) as i32
            }
        });
        map.insert("stringify".to_string().to_string(), move |x| {
            (x).to_string()
        });
        map
    });
use std::collections::HashMap;
use std::sync::LazyLock;
#[derive(Debug, Clone)]
pub struct ValueError {
    message: String,
}
impl std::fmt::Display for ValueError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "value error: {}", self.message)
    }
}
impl std::error::Error for ValueError {}
impl ValueError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}
#[doc = r" Sum type for heterogeneous dictionary values(Python fidelity)"]
#[derive(Debug, Clone, PartialEq)]
pub enum DepylerValue {
    Int(i64),
    Float(f64),
    Str(String),
    Bool(bool),
    None,
    List(Vec<DepylerValue>),
    Dict(std::collections::HashMap<String, DepylerValue>),
}
impl std::fmt::Display for DepylerValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DepylerValue::Int(i) => write!(f, "{}", i),
            DepylerValue::Float(fl) => write!(f, "{}", fl),
            DepylerValue::Str(s) => write!(f, "{}", s),
            DepylerValue::Bool(b) => write!(f, "{}", b),
            DepylerValue::None => write!(f, "None"),
            DepylerValue::List(l) => write!(f, "{:?}", l),
            DepylerValue::Dict(d) => write!(f, "{:?}", d),
        }
    }
}
impl DepylerValue {
    #[doc = r" Get length of string, list, or dict"]
    pub fn len(&self) -> usize {
        match self {
            DepylerValue::Str(s) => s.len(),
            DepylerValue::List(l) => l.len(),
            DepylerValue::Dict(d) => d.len(),
            _ => 0,
        }
    }
    #[doc = r" Check if empty"]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    #[doc = r" Get chars iterator for string values"]
    pub fn chars(&self) -> std::str::Chars<'_> {
        match self {
            DepylerValue::Str(s) => s.chars(),
            _ => "".chars(),
        }
    }
    #[doc = r" Insert into dict(mutates self if Dict variant)"]
    pub fn insert(&mut self, key: String, value: DepylerValue) {
        if let DepylerValue::Dict(d) = self {
            d.insert(key, value);
        }
    }
    #[doc = r" Get value from dict by key"]
    pub fn get(&self, key: &str) -> Option<&DepylerValue> {
        if let DepylerValue::Dict(d) = self {
            d.get(key)
        } else {
            Option::None
        }
    }
    #[doc = r" Check if dict contains key"]
    pub fn contains_key(&self, key: &str) -> bool {
        if let DepylerValue::Dict(d) = self {
            d.contains_key(key)
        } else {
            false
        }
    }
    #[doc = r" Convert to String"]
    pub fn to_string(&self) -> String {
        match self {
            DepylerValue::Str(s) => s.clone(),
            DepylerValue::Int(i) => i.to_string(),
            DepylerValue::Float(fl) => fl.to_string(),
            DepylerValue::Bool(b) => b.to_string(),
            DepylerValue::None => "None".to_string(),
            DepylerValue::List(l) => format!("{:?}", l),
            DepylerValue::Dict(d) => format!("{:?}", d),
        }
    }
    #[doc = r" Convert to i64"]
    pub fn to_i64(&self) -> i64 {
        match self {
            DepylerValue::Int(i) => *i,
            DepylerValue::Float(fl) => *fl as i64,
            DepylerValue::Bool(b) => {
                if *b {
                    1
                } else {
                    0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0),
            _ => 0,
        }
    }
    #[doc = r" Convert to f64"]
    pub fn to_f64(&self) -> f64 {
        match self {
            DepylerValue::Float(fl) => *fl,
            DepylerValue::Int(i) => *i as f64,
            DepylerValue::Bool(b) => {
                if *b {
                    1.0
                } else {
                    0.0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0.0),
            _ => 0.0,
        }
    }
    #[doc = r" Convert to bool"]
    pub fn to_bool(&self) -> bool {
        match self {
            DepylerValue::Bool(b) => *b,
            DepylerValue::Int(i) => *i != 0,
            DepylerValue::Float(fl) => *fl != 0.0,
            DepylerValue::Str(s) => !s.is_empty(),
            DepylerValue::List(l) => !l.is_empty(),
            DepylerValue::Dict(d) => !d.is_empty(),
            DepylerValue::None => false,
        }
    }
}
impl std::ops::Index<usize> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, idx: usize) -> &Self::Output {
        match self {
            DepylerValue::List(l) => &l[idx],
            _ => panic!("Cannot index non-list DepylerValue"),
        }
    }
}
impl std::ops::Index<&str> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, key: &str) -> &Self::Output {
        match self {
            DepylerValue::Dict(d) => d.get(key).unwrap_or(&DepylerValue::None),
            _ => panic!("Cannot index non-dict DepylerValue with string key"),
        }
    }
}
#[derive(Debug, Clone)]
pub struct User {
    pub name: String,
    pub age: i32,
    pub created_at: (),
}
impl User {
    pub fn new(name: String, age: i32) -> Self {
        Self {
            name,
            age,
            created_at: Default::default(),
        }
    }
    pub fn greet(&self) -> String {
        return format!("Hello, I'm {}!", self.name.clone());
    }
    pub fn is_adult(&self) -> bool {
        return self.age.clone() >= 18;
    }
}
#[derive(Debug, Clone)]
pub struct AdminUser {
    pub permissions: Vec<String>,
}
impl AdminUser {
    pub fn new(_name: String, _age: i32, permissions: Vec<String>) -> Self {
        Self { permissions }
    }
    pub fn has_permission(&self, permission: String) -> bool {
        return self.permissions.clone().contains_key(&permission);
    }
    pub fn greet(&self) -> String {
        return format!("Hello, I'm Admin {}!", self.name.clone());
    }
}
#[derive(Debug, Clone)]
pub struct FileManager {
    pub filename: String,
    pub mode: String,
    pub file: (),
}
impl FileManager {
    pub fn new(filename: String, mode: String) -> Self {
        Self {
            filename,
            mode,
            file: Default::default(),
        }
    }
    pub fn __enter__(&mut self) {
        self.file = std::fs::File::open(&self.filename.clone()).unwrap();
        return self.file.clone();
    }
    pub fn __exit__(&self, _exc_type: (), _exc_val: (), _exc_tb: ()) {
        if self.file.clone() {
            self.file.clone().close();
        };
    }
}
#[doc = "Calculate the nth Fibonacci number.\n    \n    The Fibonacci sequence is defined as:\n    - F(0) = 0\n    - F(1) = 1\n    - F(n) = F(n-1) + F(n-2) for n>1\n    \n    Args:\n        n: The position in the Fibonacci sequence\n        \n    Returns:\n        The nth Fibonacci number\n        \n    Raises:\n        ValueError: If n is negative\n        \n    Examples:\n       >>>calculate_fibonacci(0)\n        0\n       >>>calculate_fibonacci(1)\n        1\n       >>>calculate_fibonacci(10)\n        55\n    "]
#[doc = " Depyler: proven to terminate"]
pub fn calculate_fibonacci(n: i32) -> Result<i32, Box<dyn std::error::Error>> {
    let mut curr: i32 = Default::default();
    let _cse_temp_0 = n < 0;
    if _cse_temp_0 {
        return Err(Box::new(ValueError::new(
            "n must be non-negative".to_string(),
        )));
    }
    let _cse_temp_1 = n <= 1;
    if _cse_temp_1 {
        return Ok(n);
    }
    let (mut prev, mut curr) = (0, 1);
    for __sanitized in (2)..(n + 1) {
        (prev, curr) = (curr, prev + curr);
    }
    Ok(curr)
}
#[doc = "Process and categorize users.\n    \n    Args:\n        users: List of User objects to process\n        filter_adults: Whether to filter only adult users\n        \n    Returns:\n        Dictionary with 'adults' and 'minors' keys\n    "]
#[doc = " Depyler: verified panic-free"]
pub fn process_users(users: &Vec<User>, _filter_adults: bool) -> HashMap<String, Vec<User>> {
    let result = {
        let mut map = HashMap::new();
        map.insert("adults".to_string(), vec![]);
        map.insert("minors".to_string(), vec![]);
        map
    };
    for user in users.iter().cloned() {
        if user.is_adult() {
            result.get("adults").cloned().unwrap_or_default().push(user);
        } else {
            result.get("minors").cloned().unwrap_or_default().push(user);
        }
    }
    if DEBUG_MODE {
        println!("{}", format!("Processed {} users", users.len() as i32));
    }
    result
}
#[doc = "Create a message handler with a prefix."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn create_handler(prefix: String) -> Box<dyn Fn(String) -> String> {
    let handler = move |message: &str| -> String {
        return format!("{}: {}", prefix, message);
    };
    Box::new(handler)
}
#[doc = "Validate if age is in acceptable range."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn validate_age(age: i32) -> bool {
    let MIN_AGE = 0;
    let MAX_AGE = 150;
    let _cse_temp_0 = age < MIN_AGE;
    if _cse_temp_0 {
        return false;
    }
    let _cse_temp_1 = age > MAX_AGE;
    if _cse_temp_1 {
        return false;
    }
    age >= 0
}
#[doc = "Function that might trigger LSP diagnostics."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn problematic_function() -> String {
    let result: i32 = "not an int";
    result.to_string()
}
#[doc = "Process numbers using various transformations."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn process_numbers(numbers: &Vec<i32>) -> Vec<i32> {
    let doubled = numbers
        .as_slice()
        .iter()
        .cloned()
        .map(|n| (data_processors.get("double").cloned().unwrap_or_default())(n))
        .collect::<Vec<_>>();
    let _cse_temp_0 = numbers
        .as_slice()
        .iter()
        .cloned()
        .filter(|n| n > 0)
        .map(|n| {
            if 2 >= 0 && (2 as i64) <= (u32::MAX as i64) {
                ({ n } as i32)
                    .checked_pow({ 2 } as u32)
                    .expect("Power operation overflowed")
            } else {
                ({ n } as f64).powf({ 2 } as f64) as i32
            }
        })
        .sum::<i32>();
    let sum_of_squares = _cse_temp_0;
    doubled
}
#[doc = "Decorator to log function calls."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn log_calls(func: DepylerValue) -> Box<dyn Fn(()) -> ()> {
    let wrapper = move |args: ()| {
        println!("{}", format!("Calling {}", func.__name__));
        return func(args);
    };
    Box::new(wrapper)
}
#[doc = "An important operation that should be logged."]
#[doc = " Depyler: verified panic-free"]
#[doc = " Depyler: proven to terminate"]
pub fn important_operation(value: &str) -> String {
    value.to_uppercase()
}
#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{quickcheck, TestResult};
    #[test]
    fn test_calculate_fibonacci_examples() {
        assert_eq!(calculate_fibonacci(0), 0);
        assert_eq!(calculate_fibonacci(1), 1);
        assert_eq!(calculate_fibonacci(-1), -1);
    }
    #[test]
    fn test_validate_age_examples() {
        let _ = validate_age(Default::default());
    }
    #[test]
    fn test_process_numbers_examples() {
        assert_eq!(process_numbers(vec![]), vec![]);
        assert_eq!(process_numbers(vec![1]), vec![1]);
    }
}
